# Data Cleaning

### In version 2 of data cleaning, we want to introduce new columns: `home_age`, `baths`, and `garage_age`

-----
## The main objective of this notebook is to clean the data in order to have a complete dataframe (no null values). The following will be executed:
- Remove columns
- Impute missing values 
- Drop empty rows

In [1]:
import pandas as pd

In [2]:
train = pd.read_csv('./datasets/train.csv')

In [3]:
train.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  109  533352170           60        RL           NaN     13517   Pave   NaN   
1  544  531379050           60        RL          43.0     11492   Pave   NaN   
2  153  535304180           20        RL          68.0      7922   Pave   NaN   
3  318  916386060           60        RL          73.0      9802   Pave   NaN   
4  255  906425045           50        RL          82.0     14235   Pave   NaN   

  Lot Shape Land Contour    ...     Screen Porch Pool Area Pool QC Fence  \
0       IR1          Lvl    ...                0         0     NaN   NaN   
1       IR1          Lvl    ...                0         0     NaN   NaN   
2       Reg          Lvl    ...                0         0     NaN   NaN   
3       Reg          Lvl    ...                0         0     NaN   NaN   
4       IR1          Lvl    ...                0         0     NaN   NaN   

  Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  SalePrice  
0          NaN        0       3    2010        WD      130500  
1          NaN        0       4    2009        WD      220000  
2          NaN        0       1    2010        WD      109000  
3          NaN        0       4    2010        WD      174000  
4          NaN        0       3    2010        WD      138500  

[5 rows x 81 columns]

In [4]:
train.shape

(2051, 81)

In [5]:
train.isnull().mean().sort_values(ascending = False).head(7)

Pool QC          0.995612
Misc Feature     0.968308
Alley            0.931741
Fence            0.804973
Fireplace Qu     0.487567
Lot Frontage     0.160897
Garage Finish    0.055583
dtype: float64

# Removing Columns

In [6]:
#the columns I want to remove are in this list 
del_col = ['Pool QC',          
            'Misc Feature',     
            'Alley',            
            'Fence'  ]

train.drop(labels = del_col, axis = 1, inplace = True)

In [7]:
train.shape

#4 columns have been removed

(2051, 77)

In [8]:
train.isnull().mean().sort_values(ascending= False).head(25)

Fireplace Qu      0.487567
Lot Frontage      0.160897
Garage Yr Blt     0.055583
Garage Finish     0.055583
Garage Qual       0.055583
Garage Cond       0.055583
Garage Type       0.055095
Bsmt Exposure     0.028279
BsmtFin Type 2    0.027304
Bsmt Qual         0.026816
BsmtFin Type 1    0.026816
Bsmt Cond         0.026816
Mas Vnr Area      0.010726
Mas Vnr Type      0.010726
Bsmt Full Bath    0.000975
Bsmt Half Bath    0.000975
BsmtFin SF 2      0.000488
BsmtFin SF 1      0.000488
Bsmt Unf SF       0.000488
Total Bsmt SF     0.000488
Garage Cars       0.000488
Garage Area       0.000488
Land Slope        0.000000
Exterior 2nd      0.000000
PID               0.000000
dtype: float64

# Before I drop rows where cells are blank, I may want to impute missing values 

In [9]:
train.shape

(2051, 77)

# Lets look at the unique values for the columns where cell are missing

### Here are the columns where cells are missing:
    - Fireplace Qu      0.487567
    - Lot Frontage      0.160897
    - Garage Yr Blt     0.055583
    - Garage Finish     0.055583
    - Garage Qual       0.055583
    - Garage Cond       0.055583
    - Garage Type       0.055095
    - Bsmt Exposure     0.028279
    - BsmtFin Type 2    0.027304
    - Bsmt Qual         0.026816
    - BsmtFin Type 1    0.026816
    - Bsmt Cond         0.026816
    - Mas Vnr Area      0.010726
    - Mas Vnr Type      0.010726
    - Bsmt Full Bath    0.000975
    - Bsmt Half Bath    0.000975
    - BsmtFin SF 2      0.000488
    - BsmtFin SF 1      0.000488
    - Bsmt Unf SF       0.000488
    - Total Bsmt SF     0.000488
    - Garage Cars       0.000488
    - Garage Area       0.000488

# Fireplace Qu

In [10]:
train['Fireplace Qu'].value_counts()

### Value types for Fireplace Qu from http://jse.amstat.org/v19n3/decock/DataDocumentation.txt

#Ex	Excellent - Exceptional Masonry Fireplace
#Gd	Good - Masonry Fireplace in main level
#TA	Average - Prefabricated Fireplace in main living area or Masonry Fireplace in basement
#Fa	Fair - Prefabricated Fireplace in basement
#Po	Poor - Ben Franklin Stove
#NA	No Fireplace----- This part is missing!

Gd    523
TA    407
Fa     59
Po     31
Ex     31
Name: Fireplace Qu, dtype: int64

In [11]:
# changing null cells with NA for fireplace qu
train['Fireplace Qu'].fillna('none',  inplace = True)

In [12]:
train['Fireplace Qu'].value_counts()

none    1000
Gd       523
TA       407
Fa        59
Po        31
Ex        31
Name: Fireplace Qu, dtype: int64

# Lot Frontage

In [13]:
train['Lot Frontage'].value_counts().head()

#lot frontage is numeric, how should I fill empty cells when column is numeric?

60.0    179
70.0     96
80.0     94
50.0     90
65.0     71
Name: Lot Frontage, dtype: int64

In [14]:
train['Lot Frontage'].fillna(train['Lot Frontage'].median(), inplace = True)

In [15]:
train['Lot Frontage'].isnull().sum()

0

# Garage Yr Blt

In [16]:
#numeric
train['Garage Yr Blt'].head()

0    1976.0
1    1997.0
2    1953.0
3    2007.0
4    1957.0
Name: Garage Yr Blt, dtype: float64

In [17]:
train['Garage Yr Blt'].fillna(train['Garage Yr Blt'].median(), inplace = True)

In [18]:
train['Garage Yr Blt'].isnull().sum()

0

# Garage Finish

In [19]:
train['Garage Finish'].value_counts()


#Fin	Finished
#RFn	Rough Finished	
#Unf	Unfinished
#NA	No Garage

Unf    849
RFn    579
Fin    509
Name: Garage Finish, dtype: int64

In [20]:
train['Garage Finish'].fillna('none', inplace = True)

In [21]:
train['Garage Finish'].isnull().sum()

0

# Garage Qual

In [22]:
train['Garage Qual'].value_counts()

#  Ex	Excellent
#        Gd	Good
#        TA	Typical/Average
#        Fa	Fair
#        Po	Poor
#        NA	No Garage

TA    1832
Fa      82
Gd      18
Ex       3
Po       2
Name: Garage Qual, dtype: int64

In [23]:
train['Garage Qual'].fillna('none', inplace = True)

In [24]:
train['Garage Qual'].isnull().sum()

0

# Garage Cond

In [25]:
train['Garage Cond'].value_counts()

# Ex	Excellent
#        Gd	Good
#        TA	Typical/Average
#        Fa	Fair
#        Po	Poor
#        NA	No Garage

TA    1868
Fa      47
Gd      12
Po       8
Ex       2
Name: Garage Cond, dtype: int64

In [26]:
train['Garage Cond'].fillna('none', inplace = True)

In [27]:
train['Garage Cond'].isnull().sum()

0

# Garage Type

In [28]:
train['Garage Type'].value_counts()

#  2Types	More than one type of garage
#        Attchd	Attached to home
#        Basment	Basement Garage
#        BuiltIn	Built-In (Garage part of house - typically has room above garage)
#        CarPort	Car Port
#        Detchd	Detached from home
#        NA	No Garage

Attchd     1213
Detchd      536
BuiltIn     132
Basment      27
2Types       19
CarPort      11
Name: Garage Type, dtype: int64

In [29]:
train['Garage Type'].fillna('none', inplace = True)

In [30]:
train['Garage Type'].isnull().sum()

0

# Bsmt Exposure

In [31]:
train['Bsmt Exposure'].value_counts()

# Gd	Good Exposure
#        Av	Average Exposure (split levels or foyers typically score average or above)	
#        Mn	Mimimum Exposure
#        No	No Exposure
#        NA	No Basement

No    1339
Av     288
Gd     203
Mn     163
Name: Bsmt Exposure, dtype: int64

In [32]:
train['Bsmt Exposure'].fillna('none', inplace = True)

In [33]:
train['Bsmt Exposure'].isnull().sum()

0

# BsmtFin Type 2

In [34]:
train['BsmtFin Type 2'].value_counts()

# GLQ	Good Living Quarters
#        ALQ	Average Living Quarters
#        BLQ	Below Average Living Quarters	
#        Rec	Average Rec Room
#        LwQ	Low Quality
#        Unf	Unfinshed
#        NA	No Basement

Unf    1749
Rec      80
LwQ      60
BLQ      48
ALQ      35
GLQ      23
Name: BsmtFin Type 2, dtype: int64

In [35]:
train['BsmtFin Type 2'].fillna('none', inplace = True)

In [36]:
train['BsmtFin Type 2'].isnull().sum()

0

# Bsmt Qual

In [37]:
train['Bsmt Qual'].value_counts()


# Ex	Excellent (100+ inches)	
#        Gd	Good (90-99 inches)
#        TA	Typical (80-89 inches)
#        Fa	Fair (70-79 inches)
#        Po	Poor (<70 inches
#        NA	No Basement

TA    887
Gd    864
Ex    184
Fa     60
Po      1
Name: Bsmt Qual, dtype: int64

In [38]:
train['Bsmt Qual'].fillna('none', inplace = True)

In [39]:
train['Bsmt Qual'].isnull().sum()

0

## At this point, this is too tedious. I realized the missing values for object columns are NA. I'm going to create a for loop that will fill missing values for me. If the column is numeric, it will fill in the data using the median. 

In [40]:
#my list of columns that need to be filled 
fil_col = ['Bsmt Cond',         
'Mas Vnr Area',      
'Mas Vnr Type',      
'Bsmt Full Bath',    
'Bsmt Half Bath',    
'BsmtFin SF 2',      
'BsmtFin SF 1',      
'Bsmt Unf SF',       
'Total Bsmt SF',     
'Garage Cars',       
'Garage Area',
'BsmtFin Type 1']

In [41]:
for col in list(fil_col):
    if train[col].dtype != 'O':
        train[col].fillna(train[col].median(), inplace = True)
    else: 
        train[col].fillna('none', inplace = True)

In [42]:
#check if it worked 
train.isnull().sum().sum()

0

In [43]:
train.shape

(2051, 77)

# Before we save this data, let's introduce the three new columns: `home_age`, `baths`, and `garage_age`

In [44]:
#creating home_age 
train['homeage'] = train['Yr Sold'] - train['Year Built']

#creating baths
train['baths'] = train['Full Bath'] + (train['Half Bath']/2)

#creating garage_age 
train['garageage'] = train['Yr Sold'] - train['Garage Yr Blt']

In [45]:
train.shape

(2051, 80)

In [46]:
train.head(1)

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0  109  533352170           60        RL          68.0     13517   Pave   

  Lot Shape Land Contour Utilities    ...     Screen Porch Pool Area Misc Val  \
0       IR1          Lvl    AllPub    ...                0         0        0   

  Mo Sold Yr Sold Sale Type SalePrice  homeage  baths  garageage  
0       3    2010       WD     130500       34    2.5       34.0  

[1 rows x 80 columns]

In [47]:
train.isnull().sum().sum()

0

# Lets save this data and call it clean_train.csv and store it in the datasets folder

In [48]:
train.to_csv('./datasets/clean_train.csv')